File: src/main/java/cn/academy/vanilla/meltdowner/skill/JetEngine.java
/**
* Copyright (c) Lambda Innovation, 2013-2016
* This file is part of the AcademyCraft mod.
* https://github.com/LambdaInnovation/AcademyCraft
* Licensed under GPLv3, see project root for more information.
*/
package cn.academy.vanilla.meltdowner.skill;

import cn.academy.ability.api.Skill;
import cn.academy.ability.api.ctrl.ActionManager;
import cn.academy.ability.api.ctrl.SkillInstance;
import cn.academy.ability.api.ctrl.action.SkillSyncAction;
import cn.academy.ability.api.data.AbilityData;
import cn.academy.vanilla.generic.entity.EntityRippleMark;
import cn.academy.vanilla.meltdowner.client.render.MdParticleFactory;
import cn.academy.vanilla.meltdowner.entity.EntityDiamondShield;
import cn.lambdalib.annoreg.core.Registrant;
import cn.lambdalib.networkcall.RegNetworkCall;
import cn.lambdalib.networkcall.s11n.StorageOption.Data;
import cn.lambdalib.networkcall.s11n.StorageOption.Target;
import cn.lambdalib.particle.Particle;
import cn.lambdalib.util.generic.RandUtils;
import cn.lambdalib.util.generic.VecUtils;
import cn.lambdalib.util.mc.EntitySelectors;
import cn.lambdalib.util.mc.Raytrace;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.MovingObjectPosition;
import net.minecraft.util.Vec3;

import static cn.lambdalib.util.generic.MathUtils.*;

/**
 * @author WeAthFolD
 */
@Registrant
public class JetEngine extends Skill {

    public static final JetEngine instance = new JetEngine();

    private JetEngine() {
        super("jet_engine", 4);
    }
    
    @Override
    public SkillInstance createSkillInstance(EntityPlayer player) {
        SkillInstance ret = new SkillInstance().addChild(new JEAction());
        ret.estimatedCP = getConsumption(AbilityData.get(player).getSkillExp(this));
        return ret;
    }

    private static float getConsumption(float exp) {
        return lerpf(170, 140, exp);
    }
    
    public static class JEAction extends SkillSyncAction {

        float exp;
        float consumption, overload;

        public JEAction() {
            super(-1);
        }
        
        @Override
        public void onStart() {
            super.onStart();

            exp = aData.getSkillExp(instance);
            consumption = getConsumption(exp);
            overload = lerpf(66, 42, exp);
            
            if(isRemote)
                startEffects();
        }
        
        @Override
        public void onTick() {
            if(isRemote)
                updateEffects();
            
            if(!cpData.canPerform(getConsumption(exp)))
                ActionManager.abortAction(this);
        }
        
        @Override
        public void onEnd() {
            if(cpData.perform(getConsumption(exp), overload)) {
                if(!isRemote) {
                    startTriggerAction(player, getDest().addVector(0, 1.65, 0));
                    aData.addSkillExp(instance, .004f);
                    instance.triggerAchievement(player);
                }
                setCooldown(instance, (int) (18 * lerpf(6, 3, exp)));
            }
        }
        
        @Override
        public void onFinalize() {
            if(isRemote)
                endEffects();
        }
        
        Vec3 getDest() {
            return Raytrace.getLookingPos(player, 12, EntitySelectors.nothing).getLeft();
        }
        
        // CLIENT
        @SideOnly(Side.CLIENT)
        EntityRippleMark mark;
        
        @SideOnly(Side.CLIENT)
        void startEffects() {
            if(isLocal()) {
                world.spawnEntityInWorld(mark = new EntityRippleMark(world));
                mark.color.setColor4d(0.2, 1.0, 0.2, 0.7);
            }
        }
        
        @SideOnly(Side.CLIENT)
        void updateEffects() {
            if(isLocal()) {
                Vec3 dest = getDest();
                mark.setPosition(dest.xCoord, dest.yCoord, dest.zCoord);
            }
        }
        
        @SideOnly(Side.CLIENT)
        void endEffects() {
            if(isLocal()) {
                mark.setDead();
            }
        }
        
    }
    
    public static class JETriggerAction extends SkillSyncAction {
        
        static final float TIME = 8, LIFETIME = 15;
        
        Vec3 target;
        Vec3 start;
        Vec3 velocity;
        int ticks;

        float exp;

        public JETriggerAction(Vec3 _target) {
            super(-1);
            target = _target;
        }
        
        public JETriggerAction() {
            super(-1);
        }
        
        @Override
        public void writeNBTStart(NBTTagCompound tag) {
            tag.setFloat("x", (float) target.xCoord);
            tag.setFloat("y", (float) target.yCoord);
            tag.setFloat("z", (float) target.zCoord);
        }
        
        @Override
        public void readNBTStart(NBTTagCompound tag) {
            target = VecUtils.vec(tag.getFloat("x"), tag.getFloat("y"), tag.getFloat("z"));
        }
        
        @Override
        public void onStart() {
            super.onStart();

            exp = aData.getSkillExp(instance);

            if(isRemote) {
                startEffects();
                
                start = VecUtils.vec(player.posX, player.posY, player.posZ);
                velocity = VecUtils.multiply(VecUtils.subtract(target, start), 1.0 / TIME);
            }
        }
        
        @Override
        public void onTick() {
            ++ticks;
            
            if(isRemote) {
                if(ticks >= LIFETIME) {
                    ActionManager.endAction(this);
                } else if(ticks <= TIME){
                    Vec3 pos = VecUtils.lerp(start, target, ticks / TIME);
                    player.setPosition(pos.xCoord, pos.yCoord, pos.zCoord);
                    player.motionX = velocity.xCoord;
                    player.motionY = velocity.yCoord;
                    player.motionZ = velocity.zCoord;
                    player.fallDistance = 0.0f;
                }
                
                updateEffects();
            } else {
                MovingObjectPosition pos = Raytrace.perform(world, 
                    VecUtils.vec(player.lastTickPosX, player.lastTickPosY, player.lastTickPosZ),
                    VecUtils.vec(player.posX, player.posY, player.posZ),
                    EntitySelectors.and(EntitySelectors.excludeOf(player), EntitySelectors.living)
                );
                if(pos != null && pos.entityHit != null) {
                    MDDamageHelper.attack(pos.entityHit, player, lerpf(15, 35, exp));
                }
            }
        }
        
        @Override
        public void onFinalize() {
            if(isRemote)
                endEffects();
        }
        
        // CLIENT
        @SideOnly(Side.CLIENT)
        EntityDiamondShield entity;
        
        @SideOnly(Side.CLIENT)
        private void startEffects() {
            world.spawnEntityInWorld(entity = new EntityDiamondShield(player));
        }
        
        @SideOnly(Side.CLIENT)
        private void updateEffects() {
            if(isLocal()) {
                player.capabilities.setPlayerWalkSpeed(0.07f);
            }
            
            {
                for(int i = 0; i < 10; ++i) {
                    Vec3 pos2 = VecUtils.lerp(start, target, 3 * ticks / TIME);
                    Particle p = MdParticleFactory.INSTANCE.next(world,
                        VecUtils.add(VecUtils.vec(player.posX, player.posY, player.posZ), 
                        VecUtils.vec(
                            RandUtils.ranged(-.3, .3), 
                            RandUtils.ranged(-.3, .3),
                            RandUtils.ranged(-.3, .3))), 
                        VecUtils.vec(
                            RandUtils.ranged(-.02, .02), 
                            RandUtils.ranged(-.02, .02),
                            RandUtils.ranged(-.02, .02)));
                    world.spawnEntityInWorld(p);
                }
            }
        }
        
        @SideOnly(Side.CLIENT)
        private void endEffects() {
            if(isLocal()) {
                player.capabilities.setPlayerWalkSpeed(0.1f);
            }
            entity.setDead();
        }
        
    }
    
    @RegNetworkCall(side = Side.CLIENT)
    private static void startTriggerAction(@Target EntityPlayer player, @Data Vec3 vec) {
        ActionManager.startAction(new JETriggerAction(vec));
    }

}

File: src/main/scala/cn/academy/vanilla/meltdowner/skill/JetEngine.scala
/**
  * Copyright (c) Lambda Innovation, 2013-2016
  * This file is part of the AcademyCraft mod.
  * https://github.com/LambdaInnovation/AcademyCraft
  * Licensed under GPLv3, see project root for more information.
  */
package cn.academy.vanilla.meltdowner.skill

import cn.academy.ability.api.Skill
import cn.academy.ability.api.context.{ClientContext, ClientRuntime, Context, RegClientContext}
import cn.academy.vanilla.generic.entity.EntityRippleMark
import cn.academy.vanilla.meltdowner.client.render.MdParticleFactory
import cn.academy.vanilla.meltdowner.entity.EntityDiamondShield
import cn.lambdalib2.annoreg.core.Registrant
import cn.lambdalib2.particle.Particle
import cn.lambdalib2.s11n.network.NetworkMessage.Listener
import cn.lambdalib2.util.generic.{RandUtils, VecUtils}
import cn.lambdalib2.util.mc.{EntitySelectors, Raytrace}
import net.minecraftforge.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.{MovingObjectPosition, Vec3}

/**
  * @author WeAthFolD, KSkun
  */
object JetEngine extends Skill("jet_engine", 4) {

  @SideOnly(Side.CLIENT)
  override def activate(rt: ClientRuntime, keyID: Int) = activateSingleKey(rt, keyID, p => new JEContext(p))

}

object JEContext {

  final val MSG_TRIGGER = "trigger"
  final val MSG_MARK_END = "mark_end"

}

import cn.academy.ability.api.AbilityAPIExt._
import cn.lambdalib2.util.generic.MathUtils._
import JEContext._

class JEContext(p: EntityPlayer) extends Context(p, JetEngine) {

  private val exp: Float = ctx.getSkillExp
  private val consumption: Float = lerpf(170, 140, exp)
  private val overload: Float = lerpf(60, 50, exp)

  @Listener(channel=MSG_TICK, side=Array(Side.SERVER))
  private def s_onTick() = if(!ctx.canConsumeCP(consumption)) terminate()

  @Listener(channel=MSG_KEYUP, side=Array(Side.CLIENT))
  private def l_onKeyUp() = sendToServer(MSG_MARK_END)

  @Listener(channel=MSG_KEYABORT, side=Array(Side.CLIENT))
  private def l_onKeyAbort() = {
    sendToClient(MSG_MARK_END)
    terminate()
  }

  @Listener(channel=MSG_MARK_END, side=Array(Side.SERVER))
  private def s_onEnd() = {
    if(ctx.consume(consumption, overload)) {
      sendToClient(MSG_MARK_END)
      sendToSelf(MSG_TRIGGER, getDest.addVector(0, 1.65, 0))
      ctx.addSkillExp(.004f)
      JetEngine.triggerAchievement(player)
      ctx.setCooldown(lerpf(60, 30, exp).toInt)
    } else {
      sendToClient(MSG_MARK_END)
      terminate()
    }
  }

  private def getDest: Vec3 = Raytrace.getLookingPos(player, 12, EntitySelectors.nothing).getLeft

  //TRIGGER
  private val TIME: Float = 8
  private val LIFETIME: Float = 15
  private var target: Vec3 = _
  private var ticks: Int = 0
  private var isTriggering = false

  private var start: Vec3 = _
  private var velocity: Vec3 = _

  @Listener(channel=MSG_TRIGGER, side=Array(Side.SERVER))
  private def s_triggerStart(_target: Vec3) = {
    target = _target
    isTriggering = true

    sendToClient(MSG_TRIGGER, target)
  }

  @Listener(channel=MSG_TRIGGER, side=Array(Side.CLIENT))
  private def c_triggerStart(_target: Vec3) = {
    if(isLocal) {
      isTriggering = true
      target = _target

      start = VecUtils.vec(player.posX, player.posY, player.posZ)
      velocity = VecUtils.multiply(VecUtils.subtract(target, start), 1.0 / TIME)
    }
  }

  @Listener(channel=MSG_TICK, side=Array(Side.SERVER))
  private def s_triggerTick() = {
    if(isTriggering) {
      val pos: MovingObjectPosition = Raytrace.perform(world,
        VecUtils.vec(player.lastTickPosX, player.lastTickPosY, player.lastTickPosZ),
        VecUtils.vec(player.posX, player.posY, player.posZ), EntitySelectors.exclude(player).and(EntitySelectors.living))
      if(player.ridingEntity!=null)player.mountEntity(null);
      if (pos != null && pos.entityHit != null) MDDamageHelper.attack(ctx, pos.entityHit, lerpf(7, 20, exp))
    }
  }

  @Listener(channel=MSG_TICK, side=Array(Side.CLIENT))
  private def c_triggerTick(): Unit = {
    if(isLocal && isTriggering) {
      if (ticks >= LIFETIME)
        terminate()
      ticks += 1
      val pos: Vec3 = VecUtils.lerp(start, target, ticks / TIME)
      player.setPosition(pos.xCoord, pos.yCoord, pos.zCoord)
      player.motionX = velocity.xCoord
      player.motionY = velocity.yCoord
      player.motionZ = velocity.zCoord
      player.fallDistance = 0.0f
    }
  }

}

@Registrant
@SideOnly(Side.CLIENT)
@RegClientContext(classOf[JEContext])
class JEContextC(par: JEContext) extends ClientContext(par) {

  private val TIME: Float = 8
  
  private var mark: EntityRippleMark = _

  private var target: Vec3 = _
  private val start: Vec3 = VecUtils.vec(player.posX, player.posY, player.posZ)
  
  private var isMarking = false
  private var ticks: Int = 0
  
  @Listener(channel=MSG_MADEALIVE, side=Array(Side.CLIENT))
  private def l_spawnMark() = {
    if(isLocal) {
      isMarking = true
      mark = new EntityRippleMark(world)
      world.spawnEntityInWorld(mark)
      mark.color.setColor4d(0.2, 1.0, 0.2, 0.7)
    }
  }
  
  @Listener(channel=MSG_TICK, side=Array(Side.CLIENT))
  private def l_updateMark() = {
    if(isLocal && isMarking) {
      val dest: Vec3 = getDest
      mark.setPosition(dest.xCoord, dest.yCoord, dest.zCoord)
    }
  }

  @Listener(channel=MSG_MARK_END, side=Array(Side.CLIENT))
  private def l_endMark() = {
    if(isLocal) {
      isMarking = false
      mark.setDead()
    }
  }
  
  private def getDest: Vec3 = Raytrace.getLookingPos(player, 12, EntitySelectors.nothing).getLeft

  //TRIGGER
  private var entity: EntityDiamondShield = _
  private var isTriggering = false

  @Listener(channel=MSG_TRIGGER, side=Array(Side.CLIENT))
  private def c_tStartEffect(_target: Vec3) = {
    target = _target
    isTriggering = true
    entity = new EntityDiamondShield(player)
    world.spawnEntityInWorld(entity)
  }

  @Listener(channel=MSG_TICK, side=Array(Side.CLIENT))
  private def c_tUpdateEffect() = {
    if(isTriggering) {
      ticks += 1
      if (isLocal) player.capabilities.setPlayerWalkSpeed(0.07f)
      for (i <- 0 to 10) {
        val pos2: Vec3 = VecUtils.lerp(start, target, 3 * ticks / TIME)
        val p: Particle = MdParticleFactory.INSTANCE.next(world, VecUtils.add(VecUtils.vec(player.posX, player.posY, player.posZ),
          VecUtils.vec(RandUtils.ranged(-.3, .3), RandUtils.ranged(-.3, .3), RandUtils.ranged(-.3, .3))),
          VecUtils.vec(RandUtils.ranged(-.02, .02), RandUtils.ranged(-.02, .02), RandUtils.ranged(-.02, .02)))
        world.spawnEntityInWorld(p)
      }
    }
  }
  
  @Listener(channel=MSG_TERMINATED, side=Array(Side.CLIENT))
  private def c_tEndEffect() = {
    if(mark != null) mark.setDead()

    if(isTriggering) {
      if (isLocal) player.capabilities.setPlayerWalkSpeed(0.1f)
      entity.setDead()
    }
    isTriggering = false
  }

}
